src/osd/ECUtil.cc

```diff
@@ -771,6 +775,43 @@ void shard_extent_map_t::pad_and_rebuild_to_page_align() {
   }
 }
 
+// statics should be initialised to zero!
+// NOTE: Forthcoming EC-fix will more generally apply this constant in EC.
+static const uint64_t EC_ALIGN_SIZE = 4096;
+static char zero_page[EC_ALIGN_SIZE];
+
+shard_extent_set_t shard_extent_map_t::get_zeros_extent_set() const {
+  shard_extent_set_t zero_set(sinfo->get_k_plus_m());
+  for (auto &&[shard, emap] : extent_maps) {
+    for (auto i = emap.begin(); i != emap.end(); ++i) {
+      uint64_t emap_offset = i.get_off();
+      uint64_t emap_length = i.get_len();
+
+      ceph_assert(emap_offset % EC_ALIGN_SIZE == 0);
+      ceph_assert(emap_length % EC_ALIGN_SIZE == 0);
+
+      bufferlist bl = i.get_val();
+      uint64_t bl_offset = 0;
+      auto j = bl.begin();
+      while (j != bl.end()) {
+        bufferptr bp = j.get_current_ptr();
+        uint64_t bp_length = bp.length();
+        ceph_assert(bp_length % EC_ALIGN_SIZE == 0);
+        char *c_str = bp.c_str();
+        for (uint64_t bp_offset = 0; bp_offset < bp_length; bp_offset += EC_ALIGN_SIZE) {
+          if (0 == memcmp(zero_page, c_str + bp_offset, EC_ALIGN_SIZE)) {
```

Comment:
We don't really need to inspect the raw memory; it should be enough to cheaply check a _flavor_ of the underlying `bp::_raw` (or even location of its psychical zero block if it's a singleton) assuming `bufferlist`'s methods producing zeros (`append_zero()` and `prepend_zero()`) mark them accordingly. This can be nicely combined with `mprotect` to provide protection at run-time and bigger rework of the bufferlist  (`bufferlist` as acceptor of every `bptr` but donor only of its read-only variant) to assure safety at compile-time.